tools/autograd/templates/Functions.h

```diff
@@ -25,19 +25,19 @@ using at::ScalarType;
 using c10::optional;
 using c10::fmap;
 
-inline std::vector<Tensor> unpack_list(at::ArrayRef<SavedVariable> xs) {
+inline std::vector<Tensor> unpack_list(at::ArrayRef<SavedVariable> xs, std::shared_ptr<Node> saved_for = nullptr) {
   // NB: we must explicitly do the conversion in the lambda, otherwise template
   // deduction will give a Tensor of Variable which is not convertible
-  return fmap(xs, [](const SavedVariable& x) {
-    return static_cast<Tensor>(x.unpack());
+  return fmap(xs, [&saved_for](const SavedVariable& x) {
+    return static_cast<Tensor>(x.unpack(saved_for));
   });
 }
 
-inline c10::List<c10::optional<Tensor>> unpack_opt_list(at::ArrayRef<SavedVariable> xs) {
+inline c10::List<c10::optional<Tensor>> unpack_opt_list(at::ArrayRef<SavedVariable> xs, std::shared_ptr<Node> saved_for = nullptr) {
   torch::List<c10::optional<Tensor>> result;
   result.reserve(xs.size());
   for (const SavedVariable& v : xs) {
-    auto var = v.unpack();
+    auto var = v.unpack(saved_for);
```

Comment:
Oh whoops, `std::move` does not work because we would be doing it multiple times. 
It really sucks that we need to increment refcount for each input/output, but it would take a larger refactor to do this (no action necessary).